### PR TITLE
Make CMS -sign and -verify handle binary input correctly

### LINKED ISSUE #8940: `openssl cms -sign/-verify` do not handle binary input properly (including DER)
The binary input is copied to signedData.encapContentInfo.eContent only up to the first octet equal to zero, the rest gets lost. The `-binary` flag has no effect on this.

Sample command line:

    ```
    $ openssl version
    OpenSSL 1.0.2r  26 Feb 2019
    $ alias _dd="dd if=/dev/random count=1 2>/dev/null"
    $ alias _sign="openssl cms -sign -md sha256 -signer cert.pem -inkey key.pem"
    $ alias _verify="openssl cms -verify -CAfile ca.pem 2>/dev/null"
    $ for _i in {1..3}; do _dd bs=512 | _sign | _verify | wc -c; done
         351
          36
         244
    $
    ```

### PR DESCRIPTION
**Pull request by DDvO**
So far signing and verifying S/MIME binary input did not work properly.
* `openssl cms -sign` tended to corrupt/truncate non-detached binary data without warning
* `openssl cms -verify -binary` tended to give unsuccessful verification result or fail with
```
Error reading S/MIME message
139893396579456:error:0D0D40D2:asn1 encoding routines:SMIME_read_ASN1:no multipart body failure:../crypto/asn1/asn_mime.c:408:
```
This PR fixes that.
* Warn when the `-in` or `-content` argument files contains NUL or 8-bit chars without the `-binary` option being used.
* Correct parsing of non-detached binary S/MIME data such that parse failures and verification false negatives are avoided.

This fixes #8940.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: apps/cms.c [hunk 1]
```diff
@@ -266,6 +266,34 @@ static CMS_ContentInfo *load_content_info(int informat, BIO *in, BIO **indata,
     return NULL;
 }
 
+static void warn_binary(const char *file, int flags)
+{
+    BIO *bio;
+    unsigned char linebuf[1024], *cur, *end;
+    int len;
+
+    if ((flags & CMS_BINARY) != 0)
```
Comment (slontis) [inline apps/cms.c:275]: Do you mean == 0 here?

Comment (slontis) [reply apps/cms.c:275]: This logic look like if its not in binary form load the file in binary form??
May need a comment if this is right..

File: apps/cms.c [hunk 2]
```diff
@@ -869,16 +896,20 @@ int cms_main(int argc, char **argv)
             goto end;
     }
 
-    in = bio_open_default(infile, 'r', informat);
+    warn_binary(infile, flags);
```
Comment (slontis) [inline apps/cms.c:899]: maybe this should be optional?

File: doc/man3/SMIME_read_ASN1.pod [hunk 3]
```diff
@@ -9,15 +9,20 @@ SMIME_read_ASN1_ex, SMIME_read_ASN1
...
 
 =head1 DESCRIPTION
 
 SMIME_read_ASN1_ex() parses a message in S/MIME format.
 
-I<in> is a BIO to read the message from. I<x> can be used to optionally supply
+I<in> is a BIO to read the message from.
+If the I<flags> argument contains B<CMS_BINARY> the input is assumed to be in
+binary format and is not translated to canonical form.
+If in addition B<SMIME_ASCIICRLF> is set then that the binary input is assumed
```
Comment (slontis) [inline doc/man3/SMIME_read_ASN1.pod:23]: then the

File: doc/man3/SMIME_read_CMS.pod [hunk 4]
```diff
@@ -24,9 +24,13 @@ written to B<*bcont>, otherwise B<*bcont> is set to NULL.
...
 error occurred.
 
-SMIME_read_CMS_ex() is similar to SMIME_read_CMS() but can optionally supply a
-previously created I<cms> CMS_ContentInfo object. If I<cms> is NULL then it is
-identical to SMIME_read_CMS().
+SMIME_read_CMS_ex() is similar to SMIME_read_CMS() but a previously created
+I<cms> CMS_ContentInfo object can be supplied as well as some I<flags>.
+If the I<flags> argument contains B<CMS_BINARY> the input is assumed to be in
+binary format and is not translated to canonical form.
+If in addition B<SMIME_ASCIICRLF> is set then that the binary input is assumed
```
Comment (slontis) [inline doc/man3/SMIME_read_CMS.pod:31]: then the

File: crypto/asn1/asn_mime.c [hunk 5]
```diff
@@ -627,8 +627,12 @@ static int multi_split(BIO *bio, const char *bound, STACK_OF(BIO) **ret)
                 if (bpart == NULL)
                     return 0;
                 BIO_set_mem_eof_return(bpart, 0);
-            } else if (eol)
-                BIO_write(bpart, "\r\n", 2);
+            } else if (eol) {
+                if ((flags & CMS_BINARY) == 0 || (flags & SMIME_ASCIICRLF) != 0)
```
Comment (slontis) [inline crypto/asn1/asn_mime.c:631]: Are any of these paths related to the SMIME_ASCIICRLF logic being hit?

File: apps/cms.c (hunk 1 above)
Comment (slontis) [reply apps/cms.c:275]: Ahh I understand after reading further - it needs a comment.

Comment (DDvO) [reply apps/cms.c:275]: I did mean that no warning is shown if `-binary` is given - see also the warning texts :)

File: apps/cms.c (hunk 2 above)
Comment (DDvO) [reply apps/cms.c:899]: I strongly recommend having this warning by default in case a user forgot to use `-binary`
because unsuspecting users would forget to activate that warning, too ;-)

File: apps/cms.c (hunk 1 above)
Comment (DDvO) [reply apps/cms.c:275]: Comment(s) added

File: doc/man3/SMIME_read_ASN1.pod (hunk 3 above)
Comment (DDvO) [reply doc/man3/SMIME_read_ASN1.pod:23]: Changed

File: doc/man3/SMIME_read_CMS.pod (hunk 4 above)
Comment (DDvO) [reply doc/man3/SMIME_read_CMS.pod:31]: Changed as well

File: crypto/asn1/asn_mime.c (hunk 5 above)
Comment (DDvO) [reply crypto/asn1/asn_mime.c]: Good question!
It turns out that I confused SMIME_ASCIICRLF and SMIME_CRLFEOL.
Extended tests confirmed that; after corrections this should be fine now.

File: crypto/asn1/asn_mime.c [hunk 6]
```diff
@@ -613,9 +613,9 @@ static int multi_split(BIO *bio, const char *bound, STACK_OF(BIO) **ret)
                 return 0;
             }
             return 1;
-        } else if (part) {
-            /* Strip CR+LF from linebuf */
-            next_eol = strip_eol(linebuf, &len, 0);
+        } else if (part != 0) {
+            /* Strip (possibly CR +) LF from linebuf */
+                next_eol = strip_eol(linebuf, &len, flags);
```
Comment (slontis) [inline crypto/asn1/asn_mime.c:618]: indent incorrect..

File: crypto/bio/bss_file.c [hunk 7]
```diff
@@ -349,8 +361,14 @@ static int file_gets(BIO *bp, char *buf, int size)
         if (!fgets(buf, size, (FILE *)bp->ptr))
             goto err;
     }
-    if (buf[0] != '\0')
-        ret = strlen(buf);
+    after = file_pos(bp);
+    if (after < before)
+        goto err;
+    if (after - before < size - 1)
```
Comment (slontis) [inline crypto/bio/bss_file.c:367]: Are you sure these 2 lines are correct? (ftell is the location in the file)
If size = 3, before = 0
then after = 3

0123

File: doc/man3/SMIME_read_ASN1.pod (hunk 3 above)
Comment (slontis) [reply doc/man3/SMIME_read_ASN1.pod:23]: It doesnt look changed here..

**Review by slontis — COMMENTED** [review]
Thanks for adding extra tests...

File: doc/man3/SMIME_read_ASN1.pod (hunk 3 above)
Comment (DDvO) [reply doc/man3/SMIME_read_ASN1.pod]: Ah, because you referred also to the line 21 I thought that you wanted to have "the input" be replaced by "then the input", which I did.
Now I see that in line 23 there is "then that the", which should be replaced by "then the" - done now.

File: doc/man3/SMIME_read_CMS.pod (hunk 4 above)
Comment (DDvO) [reply doc/man3/SMIME_read_CMS.pod]: Same further correction done here.

File: crypto/asn1/asn_mime.c (hunk 6 above)
Comment (DDvO) [reply crypto/asn1/asn_mime.c]: Corrected.
On this occasion I also corrected a couple of pre-existing coding style issues in `asn_mine.c`.

File: crypto/bio/bss_file.c (hunk 7 above)
Comment (DDvO) [reply crypto/bio/bss_file.c:367]: Yes, I am sure that the calculation is correct.
It is not hat easy to see, but the key is that fgets reads at most `size - 1` chars - see its man page:
```
fgets() reads in at most one less than size characters from stream and stores them into the  buffer  pointed  to  by  s.
```
Thus if `size` is 3 and `before` is 0 then `after` cannot be 3 but 2,
so `after - before` equals 2 equals `size - 1`.

**Comment by DDvO** [comment]
> Thanks for adding extra tests...

Thanks also for your helpful review!
All comments addressed again.

File: crypto/asn1/asn_mime.c [hunk 8]
```diff
@@ -894,17 +897,17 @@ static int mime_hdr_addparam(MIME_HEADER *mhdr, const char *name, const char *va
...
     return strcmp((*a)->name, (*b)->name);
 }
 
 static int mime_param_cmp(const MIME_PARAM *const *a,
                           const MIME_PARAM *const *b)
 {
-    if (!(*a)->param_name || !(*b)->param_name)
-        return ! !(*a)->param_name - ! !(*b)->param_name;
+    if ((*a)->param_name == NULL || (*b)->param_name == NULL)
+        return ((*a)->param_name != NULL) - ((*b)->param_name != NULL);
```
Comment (slontis) [inline crypto/asn1/asn_mime.c:910]: Ahh my favorite !! :)

File: test/recipes/80-test_cms.t [hunk 9]
```diff
@@ -807,8 +807,8 @@ subtest "CMS binary input tests\n" => sub {
     my $input = srctop_file("test", "smcont.bin");
     my $signed = "smcont.signed";
     my $verified = "smcont.verified";
-    my $cert = srctop_file("test", "certs", "p384-root.pem");
-    my $key = srctop_file("test", "certs", "p384-root-key.pem");
+    my $cert = srctop_file("test", "certs", "ee-self-signed.pem");
```
Comment (slontis) [inline test/recipes/80-test_cms.t:810]: What caused this to change?

File: crypto/asn1/asn_mime.c (hunk 8 above)
Comment (DDvO) [reply crypto/asn1/asn_mime.c:910]: :smile:

File: test/recipes/80-test_cms.t (hunk 9 above)
Comment (DDvO) [reply test/recipes/80-test_cms.t:810]: Part of the CI is a configuration where ECC is not enabled, which made the tests fail.
So I changed they key and cert to something of type RSA, which is supported by all configurations.

Comment (slontis) [reply test/recipes/80-test_cms.t:810]: Ahh ok this was only for your new test - all good.

File: crypto/bio/bss_file.c (hunk 7 above)
Comment (slontis) [reply crypto/bio/bss_file.c:367]: Maybe make this obvious, so that someone reading this code later doesn't try changing it..
/* a comment similar to above */
len = size - 1;

Comment (slontis) [reply crypto/bio/bss_file.c:367]: I wont hold it up on this.. Will let you decide..

**Review by slontis — DISMISSED** [review]
LGTM

File: crypto/bio/bss_file.c (hunk 7 above)
Comment (DDvO) [reply crypto/bio/bss_file.c:368]: Good point to add a comment. I've added:
```
/* fgets should guarantee that input length == after - before <= size - 1 */
```

Comment (DDvO) [reply crypto/bio/bss_file.c:368]: On this occasion I've squashed all fixup commits.
Thank you for approving, and so I ask you to re-approve regarding the new comment.

Comment (slontis) [reply crypto/bio/bss_file.c:368]: Still approved.

**Comment by DDvO** [comment]
Hmm, those cert file load errors in the AppVeyor VSVER=15; Configuration: shared; Platform: x64/x86 builds such as
```
#   Failed test at ..\test\recipes\25-test_eai_data.t line 46.
Error loading file ..\..\..\test\recipes\25-test_eai_data\ascii_chain.pem
380C0000:error::x509 certificate routines:X509_load_cert_crl_file_with_libctx:no certificate or crl found:..\crypto\x509\by_file.c:252:
C:\Perl\bin\perl.exe ..\..\util\wrap.pl ..\..\apps\openssl.exe verify -nameopt utf8 -no_check_time -CAfile ..\..\..\test\recipes\25-test_eai_data\ascii_chain.pem ..\..\..\test\recipes\25-test_eai_data\ascii_leaf.pem => 1
not ok 8
```
look unrelated, right?
At least the input files are non-binary.

**Comment by mattcaswell** [comment]
Seems like this is a bug fix that does not have to go in before beta 1. Taking it out of that milestone.

**Comment by DDvO** [comment]
Also the two Travis failures reported here are unrelated and will be fixed by 13022.

**Comment by DDvO** [comment]
I had to push some fixups in order to make CI pass also for VS15 shared builds,
so @slontis would be nice if you can re-approve.

**Comment by DDvO** [comment]
Sorry for the noise pushing various fixups and analytical commits forth and back, chasing deep bugs in reading and/or writing binary files on Windoze. Meanwhile I am pretty sure that all of the `cms` app, the lib implementation in `crypto/asn1/asn_mime.c`, and (the runtime lib of) Windows itself have bugs contributing to this unfortunate situation.
I also have the strong impression that the issue reported and discussed in #1302 is (at least mostly) due to these bugs.

**Comment by DDvO** [comment]
I've meanwhile added fixes to `cms.c`, `crypto/asn1/asn_mime.c`, and `crypto/bio/bss_file.c` and it turns out that all of this is not yet sufficient. Currently it looks like that when writing binary files on Windows any LF chars are translated to CR+LF even though the file should have been openend without the flag `BIO_FP_TEXT`.
I'm going to continue on this topic next week because this week I am on vacation and will abstain from doing any OpenSSL work.

**Comment by t8m** [comment]
Is this still WIP? As it changes public API this needs to be merged before beta1 if at all.

**Comment by DDvO** [comment]
> Is this still WIP? As it changes public API this needs to be merged before beta1 if at all.

Good that you ask. BTW, by when is feature freeze for beta1 expected meanwhile?

This PR is still WIP because unfortunately I have not found the time to finish fixing the tricky bugs that surfaces for use with the Windows platform. Other things like generalizing the HTTP client have higher priority at least for me.
Still I will try to finish this or at least decouple the minor extensions done on the public API here,
such that the latter has a good chance to get into 3.0.

**Comment by DDvO** [comment]
Rebased the PR to fix merge conflicts.

I will also decouple again from this PR the cleanup of the CMS app options and doc,
which I tentatively had added here about a month ago.

**Comment by t8m** [comment]
> Good that you ask. BTW, by when is feature freeze for beta1 expected meanwhile?

The last alpha release was announced to happen on May 20th. So unless there is something necessary to fix a 3.0 blocker it should be in before that date.

**Comment by DDvO** [comment]
For now I've disabled (most of) the new tests that may still fail on Windows.
This way we can handle this PR, including the API change, already now
and take care of the remaining issue during the beta phase.

File: crypto/bio/bss_file.c [hunk 10]
```diff
@@ -346,10 +346,22 @@ static long file_ctrl(BIO *b, int cmd, long num, void *ptr)
...
+        return ftell((FILE *)bp->ptr);
+}
+
 static int file_gets(BIO *bp, char *buf, int size)
 {
     int ret = 0;
+    long before, after;
 
+    before = file_pos(bp);
+    if (before < 0)
```
Comment (t8m) [inline crypto/bio/bss_file.c:363]: This is bad idea! ftell is not guaranteed to work on file BIOs.

File: doc/man3/BIO_read.pod [hunk 11]
```diff
@@ -36,6 +36,8 @@ however; for example, BIO_gets() on a digest BIO will calculate and
 return the digest and other BIOs may not support BIO_gets() at all.
 The returned string is always NUL-terminated and the '\n' is preserved
 if present in the input data.
+On binary input there may be NUL characters within the string;
```
Comment (t8m) [inline doc/man3/BIO_read.pod:39]: I would be very hesitant changing the semantics of BIO_gets() like this. It could have serious unwanted consequences on third party code.

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/bio/bss_file.c (hunk 10 above)
Comment (DDvO) [reply crypto/bio/bss_file.c]: Oh, good to know!
Likely this was the reason for the strange test failures on Windows.

File: doc/man3/BIO_read.pod (hunk 11 above)
Comment (DDvO) [reply doc/man3/BIO_read.pod:41]: All right, reverted the changes to `BIO_gets()`. Instead, documented its bug on input containing the NUL char
and added a new independent function `BIO_get_line()` that works correctly on binary input (which include NUL).
This function is now used where needed in `asn_mime.c`: in `multi_split().`

**Comment by DDvO** [comment]
Thanks @t8m for getting back to this and your good further comments.
Today I found the time to handle them and thus come up with a more satisfactory and backward-compatible solution.

**Comment by DDvO** [comment]
I believe that meanwhile all issues handled here are fixed.
@slontis and/or @t8m, can you please have a look again?

File: doc/man3/BIO_read.pod [hunk 12]
```diff
@@ -72,19 +90,21 @@ a retry instead of blocking.
 See L<BIO_should_retry(3)> for details of how to
 determine the cause of a retry and other I/O issues.
 
-If the BIO_gets() function is not supported by a BIO then it possible to
-work around this by adding a buffering BIO L<BIO_f_buffer(3)>
-to the chain.
+If the "gets" method is not supported by a BIO then BIO_get_line() can be used.
+In this case it is also possible to use BIO_gets()
```
Comment (t8m) [inline doc/man3/BIO_read.pod:94]: This seems confusing a little. I'd use something like: `It is also possible to make BIO_gets() usable by ...`

File: doc/man3/BIO_read.pod [hunk 13]
```diff
@@ -36,6 +38,15 @@ however; for example, BIO_gets() on a digest BIO will calculate and
 return the digest and other BIOs may not support BIO_gets() at all.
 The returned string is always NUL-terminated and the '\n' is preserved
 if present in the input data.
+On binary input there may be NUL characters within the string;
+in this case the return value (if nonnegative) may give an incorrect length.
+
+BIO_get_line() attempts to read from BIO <b> a line of data up to the next '\n'
```
Comment (t8m) [inline doc/man3/BIO_read.pod:44]: I'd somehow mention that it reads byte-by-byte. So it might not be the most effective way to read a file. Unfortunately there is no way around that.

File: doc/man3/BIO_read.pod (hunk 12 above)
Comment (DDvO) [reply doc/man3/BIO_read.pod]: Yep, done.

File: doc/man3/BIO_read.pod (hunk 13 above)
Comment (DDvO) [reply doc/man3/BIO_read.pod:44]: Good point. Done like this:
`For implementing this, unfortunately the data needs to be read byte-by-byte.`

**Comment by DDvO** [comment]
Anything more to do on this PR?

File: doc/man3/BIO_read.pod [hunk 14]
```diff
@@ -91,7 +92,7 @@ See L<BIO_should_retry(3)> for details of how to
 determine the cause of a retry and other I/O issues.
 
 If the "gets" method is not supported by a BIO then BIO_get_line() can be used.
-In this case it is also possible to use BIO_gets()
+In this case it is also possible to make BIO_gets() usable
```
Comment (t8m) [inline doc/man3/BIO_read.pod:95]: This is still slightly confusing as at first sight it is not clear what the `In this case` refers to. When I read it first it was unclear tp me why I would use BIO_gets when I would use BIO_get_line(). Only then I realized it refers to the condition.

Comment (DDvO) [reply doc/man3/BIO_read.pod]: Ah, so I've made it more explicit:
```
It is also possible to make BIO_gets() usable even if the "gets" method is not
supported by adding a buffering BIO L<BIO_f_buffer(3)> to the chain.

Comment (DDvO) [reply doc/man3/BIO_read.pod]: Or would you prefer the following?
```
It is also possible to make BIO_gets() more generally usable 
by adding a buffering BIO L<BIO_f_buffer(3)> to the chain.

Comment (t8m) [reply doc/man3/BIO_read.pod]: It is fine as is now. Thanks.

File: crypto/bio/bio_lib.c [hunk 15]
```diff
@@ -478,6 +478,37 @@ int BIO_gets(BIO *b, char *buf, int size)
...
+    if (!bio->init) {
+        ERR_raise(ERR_LIB_BIO, BIO_R_UNINITIALIZED);
+        return -1;
+    }
+
+    while (size-- > 1 && (ret = BIO_read(bio, ptr, 1)) > 0)
+        if (*ptr++ == '\n')
+            break;
+    *ptr = '\0';
+    return ret <= 0 ? ret : ptr - buf;
```
Comment (t8m) [inline crypto/bio/bio_lib.c:509]: This returns `ret` even if some data was successfully read. For example if the input file is not terminated with `\n` it will return 0 for the last line even if it is not empty. I am not sure this is a good idea.

File: crypto/cms/cms_sd.c [hunk 16]
```diff
@@ -547,8 +547,11 @@ EVP_MD_CTX *CMS_SignerInfo_get0_md_ctx(CMS_SignerInfo *si)
 
 STACK_OF(CMS_SignerInfo) *CMS_get0_SignerInfos(CMS_ContentInfo *cms)
 {
-    CMS_SignedData *sd = cms_get0_signed(cms);
+    CMS_SignedData *sd;
 
+    ERR_set_mark();
```
Comment (t8m) [inline crypto/cms/cms_sd.c:552]: Why is this needed here?

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/bio/bio_lib.c (hunk 15 above)
Comment (DDvO) [reply crypto/bio/bio_lib.c:509]: Oh, I forgot to consult `BIO_eof()` in the loop above!
Good that you found that. Fixed.

File: crypto/cms/cms_sd.c (hunk 16 above)
Comment (DDvO) [reply crypto/cms/cms_sd.c:552]: This is explained in the commit message:
```
CMS_get0_SignerInfos(): Pervent spurious error on cms_get0_signed() failure
```

Comment (t8m) [reply crypto/cms/cms_sd.c:552]: Hmm, right. I did not look at the commits individually. There is also a typo in the commit message.

Did you get any insight why the spurious error appears? Would it be better to fix it at the place?

File: crypto/bio/bio_lib.c [hunk 17]
```diff
@@ -502,7 +502,7 @@ int BIO_get_line(BIO *bio, char *buf, int size)
         return -1;
     }
 
-    while (size-- > 1 && (ret = BIO_read(bio, ptr, 1)) > 0)
+    while (size-- > 1 && !BIO_eof(bio) && (ret = BIO_read(bio, ptr, 1)) > 0)
```
Comment (t8m) [inline crypto/bio/bio_lib.c:505]: I am wondering if BIO_eof() is always set immediately after the last possible read. I.E., if this will work properly on the case of no EOL at EOF.

It might be better to just check the BIO_eof() at the return. `return ret > 0 || BIO_eof(bio) ? ptr - buf : ret;`

File: crypto/cms/cms_sd.c (hunk 16 above)
Comment (DDvO) [reply crypto/cms/cms_sd.c]: I just re-discovered the stacktrace for it:
```
	cms_get0_signed() at cms_sd.c:30 0x7ffff7b83a98	
	CMS_get0_SignerInfos() at cms_sd.c:553 0x7ffff7b84fc8	
	ossl_cms_SignerInfos_set_cmsctx() at cms_sd.c:499 0x7ffff7b84e52	
	ossl_cms_resolve_libctx() at cms_lib.c:101 0x7ffff7b809a4	
	SMIME_read_CMS_ex() at cms_io.c:101 0x7ffff7b7f48d	
	load_content_info() at cms.c:242 0x4353e6	
	cms_main() at cms.c:896 0x436dce	
	do_cmd() at openssl.c:414 0x44dbb3	
	main() at openssl.c:295 0x44d7e2	
```
and found a more elegant solution: replace the error queue manipulation by
```
    if (OBJ_obj2nid(cms->contentType) != NID_pkcs7_signed)
        return NULL;
```

File: crypto/bio/bio_lib.c (hunk 17 above)
Comment (DDvO) [reply crypto/bio/bio_lib.c]: Good thought - checking `BIO_eof()` after the loop should be safer, and it's definitely more efficient!
Done.

File: crypto/cms/cms_sd.c (hunk 16 above)
Comment (DDvO) [reply crypto/cms/cms_sd.c]: > There is also a typo in the commit message.

Opos, fixed as well.

File: crypto/cms/cms_sd.c [hunk 18]
```diff
@@ -28,6 +28,7 @@ static CMS_SignedData *cms_get0_signed(CMS_ContentInfo *cms)
 {
     if (OBJ_obj2nid(cms->contentType) != NID_pkcs7_signed) {
         ERR_raise(ERR_LIB_CMS, CMS_R_CONTENT_TYPE_NOT_SIGNED_DATA);
+        exit(1);
```
Comment (t8m) [inline crypto/cms/cms_sd.c:31]: Debugging leftover.

Comment (DDvO) [reply crypto/cms/cms_sd.c]: Argh, removed.

File: crypto/cms/cms_sd.c (hunk 16 above)
Comment (t8m) [reply crypto/cms/cms_sd.c]: I would actually do it differently. Fix it in ossl_cms_SignerInfos_set_cmsctx() - just put the set_mark/pop_to_mark calls there.

Comment (t8m) [reply crypto/cms/cms_sd.c]: My reasoning - the problem is because the set_cmsctx() calls CMS_get0_SignerInfos() regardless of the contentType, not because the CMS_get0_SignerInfos() should not report the error. Or, alternatively it might make sense to drop the ERR_raise completely, but I am not sure about that.

Comment (DDvO) [reply crypto/cms/cms_sd.c]: Ok, if you prefer. Done.

Comment (DDvO) [reply crypto/cms/cms_sd.c]: Just saw your reasoning. Sounds reasonable :)

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Good work!

**Comment by DDvO** [comment]
> Good work!

Merged -
thanks @slontis and @t8m for your good comments and approvals!
